File: 백준/Silver/1874. 스택 수열/스택 수열.py
import sys
read = sys.stdin.readline

n = int(read())
l = [0] * n

for i in range(n):
    l[i] = int(read())

num = 1
s = []
answer = ''
result = True

for i in range(n):
    if l[i] >= num:
        while l[i] >= num:
            s.append(num)
            num += 1
            answer += '+\n'
        s.pop()
        answer += '-\n'
    else:
        p = s.pop()
        if p > l[i]:
            print('NO')
            result = False
            break
        else:
            answer += '-\n'

if result:
    print(answer)

File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.*;
import java.util.*;

public class Main{
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		
		int n = Integer.parseInt(br.readLine());
	
		int num = 1;
		Stack<Integer> stack = new Stack<>();
		boolean isOk = true;
		
		for(int i = 0; i < n; i++) {
			int find = Integer.parseInt(br.readLine());
			while(num <= find)
			{
				sb.append("+\n");
				stack.push(num);
				num++;
			}
			if(stack.pop() == find)
				sb.append("-\n");
			else {
				isOk = false;
			}
		}
		
		if(isOk) {
			System.out.println(sb.toString());
		}
		else
			System.out.println("NO");
	}
}